### ISSUE TITLE
TypeScript compiler incorrectly reports duplicate identifiers for types and variables

### ISSUE THREAD

**Issue by sulthonzh** [microsoft/TypeScript#63510]
🐛 **BUG REPORT: microsoft/TypeScript**

TypeScript compiler incorrectly reports duplicate identifiers for types and variables that should not be considered duplicates.

**Description:**
The TypeScript compiler incorrectly flags types and variables as duplicate identifiers when they are not actually duplicates. This occurs when a user-defined type or variable name collides with a built-in type or variable from the lib files, even when they should be treated as distinct.

**Reproduction steps:**
1. Create a TypeScript file with the following content:
```typescript
type Readonly<T> = {
    readonly [P in keyof T]: T[P];
};

type Event = string;

const event: Event = "test";
```

2. Compile with TypeScript: `tsc file.ts`

**Expected behavior:**
The code should compile successfully without any duplicate identifier errors.

**Actual behavior:**
The compiler reports multiple duplicate identifier errors:
```
error TS2300: Duplicate identifier 'Readonly'.
error TS2300: Duplicate identifier 'Event'.
error TS2451: Cannot redeclare block-scoped variable 'event'.
```

**Analysis:**
The issue appears to be in how TypeScript's name resolution handles collisions between user-defined names and built-in names from the lib files. The compiler should distinguish between:
- User-defined types and built-in types
- User-defined variables and built-in variables
- Different scopes (e.g., type declarations vs variable declarations)

**Environment:**
- TypeScript version: 6.0.0
- Node.js version: 20.5.1
- Operating System: macOS

**Additional context:**
This issue affects TypeScript's ability to work with common naming conventions and can lead to false-positive errors that confuse developers. The problem is particularly noticeable when working with generic type parameters that commonly used names like `Readonly`, `Event`, etc.

**Comment by MartinJohns** [comment]
This is working as intended. There is no built-in `Event` type. It's a type provided in a declaration file, which is indistinguishable from user files (user can provide their own declaration files). If you don't want these types, you can opt to exclude the standard declaration files (using `nolib` https://www.typescriptlang.org/tsconfig/#noLib).

**Comment by mkantor** [comment]
This only happens because your code is in a script (not a module; i.e. it's in the global scope). If you add any `import`s or `export`s or set [`"moduleDetection": "force"`](https://www.typescriptlang.org/tsconfig/#moduleDetection) in tsconfig then you won't see these errors.

**Comment by RyanCavanaugh** [comment]
Bot
